tests/flytekit/unit/core/test_type_hints.py

```diff
@@ -1427,3 +1431,177 @@ def foo3(a: typing.Dict) -> typing.Dict:
 
     with pytest.raises(TypeError, match="Not a collection type simple: STRUCT\n but got a list \\[{'hello': 2}\\]"):
         foo3(a=[{"hello": 2}])
+
+
+def test_union_type():
+    ut = typing.Union[int, str, float, FlyteFile, FlyteSchema, typing.List[int], typing.Dict[str, int]]
+
+    @task
+    def t1(a: ut) -> ut:
+        return a
+
+    @workflow
+    def wf(a: ut) -> ut:
+        return t1(a=a)
+
+    assert wf(a=2) == 2
+    assert wf(a="2") == "2"
+    assert wf(a=2.0) == 2.0
+    file = tempfile.NamedTemporaryFile()
+    assert isinstance(wf(a=FlyteFile(file.name)), FlyteFile)
+    assert isinstance(wf(a=FlyteSchema()), FlyteSchema)
+    assert wf(a=[1, 2, 3]) == [1, 2, 3]
+    assert wf(a={"a": 1}) == {"a": 1}
+
+    @task
+    def t2(a: typing.Union[float, dict]) -> typing.Union[float, dict]:
+        return a
+
+    @workflow
+    def wf2(a: typing.Union[int, str]) -> typing.Union[int, str]:
+        return t2(a=a)
+
+    with pytest.raises(
+        TypeError,
+        match=dedent(r'''
+            Cannot convert from scalar {
+              union {
+                value {
+                  scalar {
+                    primitive {
+                      string_value: "2"
+                    }
+                  }
+                }
+                type {
+                  simple: STRING
+                  structure {
+                    tag: "str"
+                  }
+                }
+              }
+            }
+             to typing.Union\[float, dict\] \(using tag str\)
+        ''')[1:-1],
+    ):
+        assert wf2(a="2") == "2"
+
+
+def test_optional_type():
+    @task
+    def t1(a: typing.Optional[int]) -> typing.Optional[int]:
+        return a
+
+    @workflow
+    def wf(a: typing.Optional[int]) -> typing.Optional[int]:
+        return t1(a=a)
+
+    assert wf(a=2) == 2
+    assert wf(a=None) is None
+
+
+def test_optional_type_implicit_wrapping():
+    @task
+    def t1(a: int) -> typing.Optional[int]:
+        return a if a > 0 else None
+
+    @workflow
+    def wf(a: int) -> typing.Optional[int]:
+        return t1(a=a)
+
+    assert wf(a=2) == 2
+    assert wf(a=-10) is None
+
+
+def test_union_type_implicit_wrapping():
+    @task
+    def t1(a: int) -> typing.Union[int, str]:
+        return a if a > 0 else str(a)
+
+    @workflow
+    def wf(a: int) -> typing.Union[int, str]:
+        return t1(a=a)
+
+    assert wf(a=2) == 2
+    assert wf(a=-10) == "-10"
+
+
+def test_union_type_ambiguity_checking():
```

Comment:
Don't entirely understand the last two tests here... not quite sure why one is ambiguous and the other is not?  Following the changes in the gist, `int` and `MyInt` are distinct Literal Types right?

Reply:
@wild-endeavor , in `test_union_type_ambiguity_checking` we don't have enough type information in the incoming literal (i.e. the `int` that's passed to `t1`), whereas in `test_union_type_ambiguity_resolution` `t1` takes a `Union[int, MyInt]`, but notice how that comes from the result of `t2`, which will produce a literal that contains the necessary type information to disambiguate.